### PR TITLE
HDS-336 Add new ButtonSet component, update component-checklist docs

### PR DESCRIPTION
**Pull request by KristinLBradley**
### :pushpin: Summary

Preliminary PR for new ButtonSet component. Please let me know if I missed anything.

I also updated the "Component Checklist" md file for devs. Let me know if I should revert & do as separate a PR instead. Also let me know if you disagree with my edits or have suggestions for different ones.

I added comments on various lines of my code with questions I had.

### :hammer_and_wrench: Detailed description

- Added new ButtonSet component
- Updated "Component Checklist" md.

### :camera_flash: Screenshots
Screen Shot 2022-07-28 at 10 37 55 AM

***

Reviewer's checklist:

- [ ] +1 Percy if applicable
- [ ] Confirm that PR has a changelog update via [Changesets](https://github.com/changesets/changesets) if needed

:speech_balloon: Please consider using [conventional comments](https://conventionalcomments.org/) when reviewing this PR.

### CODE REVIEW

File: packages/components/NEW-COMPONENT-CHECKLIST.md [hunk 1]
```diff
@@ -76,9 +76,13 @@ The engineering checklist has six parts: creating the feature branch, component
 ### Component Creation
 
 - [ ] create new branch from main for the component (`git checkout -b USER/COMPONENT-NAME`)
+  - Note: There is no standardized branch naming convention.
+  - Suggestion: Use Jira ticket number together with a brief description. (Including the ticket number may allow us to automatically link branches to tickets and PRs in the future.)
+  - Example Branch Name: "hds-123-fix-typo".
```
Comment (KristinLBradley) [inline packages/components/NEW-COMPONENT-CHECKLIST.md:81]: Let me know if this is helpful or if I should edit/remove. Or if this should be done as a separate PR.

File: packages/components/NEW-COMPONENT-CHECKLIST.md [hunk 2]
```diff
@@ -76,9 +76,13 @@ The engineering checklist has six parts: creating the feature branch, component
...
 - [ ] create new component using a blueprint
   - `ember generate hds-component COMPONENT-NAME`
   - if it's a child component, then `hds/COMPONENT_NAME/CHILD_NAME`
+  - (Note: Many of the below mentioned files will be automatically generated.)
```
Comment (KristinLBradley) [inline packages/components/NEW-COMPONENT-CHECKLIST.md:85]: Again, let me know if this ok or if I should change or move to new PR.

File: packages/components/addon/components/hds/button-set/index.hbs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+{{! TODO: Add optional tag variations? p tag, etc. vs just div? }}
```
Comment (KristinLBradley) [inline packages/components/addon/components/hds/button-set/index.hbs:1]: Is just a div ok? Ideally I would want to add options to use various tags but let me know. (I will delete this comment before merging.)

File: packages/components/addon/components/hds/button-set/index.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import Component from '@glimmer/component';
+// import { assert } from '@ember/debug';
```
Comment (KristinLBradley) [inline packages/components/addon/components/hds/button-set/index.js:2]: I can delete this commented out line if I don't need to add any asserts here.

Is the below enough or are there additional items I should add?

File: packages/components/tests/dummy/app/templates/components/button-set.hbs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    }}
+  </div>
+</section>
+
+{{! <section>
+  <h3 class="dummy-h3" id="accessibility"><a href="#accessibility" class="dummy-link-section">§</a> Accessibility</h3>
+  <p class="dummy-paragraph">
+    This component has been designed and implemented with accessibility in mind. When used as recommended, there should not be any WCAG conformance issues with this component.
+  </p>
+</section> }}
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/button-set.hbs:67]: I couldn't think of any special accessibility concerns for this component as it's so simple. Let me know if I should include some sort of mention about anything related to accessibility or else I can delete this commented out section.

File: packages/components/tests/dummy/app/templates/components/button-set.hbs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+<section data-test-percy>
+  <h3 class="dummy-h3" id="showcase"><a href="#showcase" class="dummy-link-section">§</a> Showcase</h3>
+
+  <Hds::ButtonSet>
+    <Hds::Button @text="Button 1" />
+    <Hds::Button @text="Button 2" />
+    <Hds::Button @text="Button 3" />
+  </Hds::ButtonSet>
+</section>
```
Comment (KristinLBradley) [inline packages/components/tests/dummy/app/templates/components/button-set.hbs:77]: Is this an ok example? Are there other more real world related examples you think would be helpful to add?

File: packages/components/tests/integration/components/hds/button-set/index-test.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+  test('it renders the component', async function (assert) {
+    await render(hbs`<Hds::ButtonSet />`);
+    assert.dom(this.element).exists();
+  });
+  test('it should render with a CSS class that matches the component name', async function (assert) {
+    await render(hbs`<Hds::ButtonSet id="test-button-set" />`);
+    assert.dom('#test-button-set').hasClass('hds-button-set');
+  });
+});
```
Comment (KristinLBradley) [inline packages/components/tests/integration/components/hds/button-set/index-test.js:17]: Are there any additional tests which would be helpful to add?

**Comment by KristinLBradley** [comment]
Looks like I need to add a changeset? I will read through how to add.

**Comment by didoo** [comment]
> Looks like I need to add a changeset? I will read through how to add.

@KristinLBradley yes, you have to run `yarn changeset` in the project root and follow the instructions.

(I have also updated the "New Component Checklist" to include instructions for this: https://github.com/hashicorp/design-system/pull/487

File: packages/components/addon/components/hds/button-set/index.hbs (hunk 3 above)
Comment (didoo) [reply packages/components/addon/components/hds/button-set/index.hbs:1]: At the moment we don't provide this functionality in our component (having dynamic tags in Ember is not possible, see for example what needs to be done to achieve something similar in this PR: https://github.com/hashicorp/cloud-ui/pull/2574/files#diff-aca16c63293e60d44e3b43fa1db3a55266ac4811697d1c4d20dcf35b81c62f70)

File: packages/components/addon/components/hds/button-set/index.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+export default class HdsButtonSetIndexComponent extends Component {
+  /**
+   * Get the class names to apply to the component.
+   * @method classNames
+   * @return {string} The "class" attribute to apply to the component.
+   */
+  get classNames() {
+    let classes = ['hds-button-set'];
```
Comment (didoo) [inline packages/components/addon/components/hds/button-set/index.js:11]: when there is no logic attached to the classnames generation, we write directly the class in the HBS file (see for example https://github.com/hashicorp/design-system/blob/ed96740352c8a1bda0cebbcec41eb6eb7de0a67d/packages/components/addon/components/hds/form/indicator/index.hbs#L2)

File: packages/components/addon/components/hds/button-set/index.js [hunk 9]
```diff
@@ -0,0 +1,15 @@
+import Component from '@glimmer/component';
```
Comment (didoo) [inline packages/components/addon/components/hds/button-set/index.js:1]: if you remove the `get classNames` method below and move the class directly in HBS, you can delete this backing class file.

File: packages/components/app/styles/components/button-set.scss [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+// BUTTON-SET
+//
+// properties within each class are sorted alphabetically
+//
+
+.hds-button-set {
+    display: flex;
+
+    > * + * {
+        margin-inline-start: 16px !important;
```
Comment (didoo) [inline packages/components/app/styles/components/button-set.scss:11]: [issue?] why `!important` here?

Comment (didoo) [reply packages/components/app/styles/components/button-set.scss:11]: also, what do you think of using just `margin-left` here? I know `margin-inline-start` is more correct, but the likelihood of having our apps support a different writing mode is approximate to zero :)

File: packages/components/tests/dummy/app/styles/pages/db-button-set.scss [hunk 11]
```diff
@@ -0,0 +1 @@
+// BUTTON-SET
```
Comment (didoo) [inline packages/components/tests/dummy/app/styles/pages/db-button-set.scss:1]: if we don't need it, we can remove it and remove the import from the `packages/components/tests/dummy/app/styles/app.scss` file

File: packages/components/tests/dummy/app/templates/components/button-set.hbs [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+  {{! prettier-ignore-end }}
+</section>
+
+<section>
+  <h3 class="dummy-h3" id="design-guidelines">
+    <a href="#design-guidelines" class="dummy-link-section">§</a>
+    Design guidelines
+  </h3>
+
+  <div class="dummy-design-guidelines">
```
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/button-set.hbs:47]: maybe we could add here the screenshot of the specific guideline related to the button set?
button-set

File: packages/components/tests/dummy/app/templates/components/button-set.hbs [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+  {{! prettier-ignore-start }}
+  <CodeBlock
+    @language="markup"
+    @code='
+      <Hds::ButtonSet>
+        <Hds::Button @text="Button 1" />
+        <Hds::Button @text="Button 2" />
+        <Hds::Button @text="Button 3" />
+      </Hds::ButtonSet>
```
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/button-set.hbs:35]: What do you think of having a more realistic use case here? Maybe just a primary and a secondary button?

File: packages/components/tests/dummy/app/templates/components/button-set.hbs [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+
+{{! <section>
+  <h3 class="dummy-h3" id="accessibility"><a href="#accessibility" class="dummy-link-section">§</a> Accessibility</h3>
+  <p class="dummy-paragraph">
+    This component has been designed and implemented with accessibility in mind. When used as recommended, there should not be any WCAG conformance issues with this component.
+  </p>
+</section> }}
+
+<section data-test-percy>
+  <h3 class="dummy-h3" id="showcase"><a href="#showcase" class="dummy-link-section">§</a> Showcase</h3>
```
Comment (didoo) [inline packages/components/tests/dummy/app/templates/components/button-set.hbs:70]: In _theory_ we could think of more extreme use cases (eg multiple buttons, with some of them with very long text) so we should have a `flex-wrap` and `vertical margin` + `flex none` on the children, but it's completely unrealistic so I would not bother about it.
screenshot_1708

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 6 above)
Comment (didoo) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:77]: Yes, having a more realistic use case would be better  (maybe just a primary and a secondary button?)

File: packages/components/tests/integration/components/hds/button-set/index-test.js (hunk 7 above)
Comment (didoo) [reply packages/components/tests/integration/components/hds/button-set/index-test.js:17]: If it was me, probably I would do something like this:
```
  test('it should render a child button', async function (assert) {
    await render(
      hbs`<Hds::ButtonSet id="test-button-set"><Hds::Button @text="my button" /></Hds::ButtonSet>`
    );
    assert.dom('#test-button-set .hds-button').exists();
  });
```
It may be redundant, but essentially it tests that the content is yielded.

**Review by didoo — COMMENTED** [review]
Added some comments

File: packages/components/NEW-COMPONENT-CHECKLIST.md (hunk 1 above)
Comment (MelSumner) [reply packages/components/NEW-COMPONENT-CHECKLIST.md:81]: A few considerations:
- I do `melsumner/branch-name` because it makes it easier to identify my branches (both in the branch UI and in tools like Tower)
- We have a spot in the PR description to link to the issue itself 

Either way, also agree with @didoo that this should be in a separate PR 👍

File: packages/components/addon/components/hds/button-set/index.hbs (hunk 3 above)
Comment (MelSumner) [reply packages/components/addon/components/hds/button-set/index.hbs:1]: A div is fine, we shouldn't (and don't) provide options to change the tag name. 

I think this could be a template-only component. I don't think it needs a backing JS file; since there's just one class, the class name can go into the template. I also don't think we need `...attributes` but I am open to hearing some thoughts on what some use cases might be.

File: packages/components/NEW-COMPONENT-CHECKLIST.md (hunk 1 above)
Comment (KristinLBradley) [reply packages/components/NEW-COMPONENT-CHECKLIST.md:81]: @MelSumner So would you like me to undo these changes and create a separate PR later?

File: packages/components/NEW-COMPONENT-CHECKLIST.md (hunk 2 above)
Comment (KristinLBradley) [reply packages/components/NEW-COMPONENT-CHECKLIST.md:85]: @MelSumner Would you like me to undo this change and move to a new PR?

File: packages/components/addon/components/hds/button-set/index.hbs (hunk 3 above)
Comment (KristinLBradley) [reply packages/components/addon/components/hds/button-set/index.hbs:1]: Would end users be able to add additional attributes if needed if I removed `...attributes`? For example if they wanted to add an aria-label or data attribute or something for some purpose?

File: packages/components/app/styles/components/button-set.scss (hunk 10 above)
Comment (KristinLBradley) [reply packages/components/app/styles/components/button-set.scss:11]: I thought you wanted to basically replicate this component: https://github.com/hashicorp/structure/blob/dffbb0e0ae5246d3b832586ab25bcb981a052b30/packages/pds-ember/app/styles/pds/components/button-set/__config.scss#L10-L14

That's the only reason I included !important as I thought there was a purpose behind using it in the original component. I'll remove it.

For the margin style, I would normally just use `margin-left` unless there's a reason not to but I see that logical property margin values were used elsewhere in the component code so I assumed that was preferred.

@MelSumner Do we want to use logical properties for accessibility reasons? Or should I just use `margin-left` here?

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 13 above)
Comment (KristinLBradley) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:35]: Yes, that's what I was thinking would be good. Is there a good example of a more realistic typical usage? I can look around.

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 12 above)
Comment (KristinLBradley) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:47]: Thanks, I can do that

File: packages/components/NEW-COMPONENT-CHECKLIST.md (hunk 1 above)
Comment (KristinLBradley) [reply packages/components/NEW-COMPONENT-CHECKLIST.md:81]: I'll just undo this change since it doesn't seem like it's useful currently. 

I would like to try to get the feature to link branches to tickets & PRs working eventually if we can though since I think it's very helpful to connect them automatically vs. having to add a link more manually.

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 12 above)
Comment (KristinLBradley) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:47]: @didoo Sorry, where should the screenshot image file go? I found a folder under "flight-website/public/assets/images" but that doesn't look like the right place...

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 13 above)
Comment (didoo) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:35]: The Figma docs:
screenshot_1714

One primary button, next to a secondary button.

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 12 above)
Comment (didoo) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:47]: @KristinLBradley this is the folder: `packages/components/tests/dummy/public/assets/images/`
(the "scrappy" website lives under the "test/dummy" Ember app, is a special app used for testing, but we've used it for the website too).

File: packages/components/app/styles/components/button-set.scss (hunk 10 above)
Comment (didoo) [reply packages/components/app/styles/components/button-set.scss]: Maybe let's not overthink this: we can just use `margin-left`, it works and it's clear what it does.

File: packages/components/addon/components/hds/button-set/index.hbs (hunk 3 above)
Comment (didoo) [reply packages/components/addon/components/hds/button-set/index.hbs]: Agree with @KristinLBradley here: we tend to add `...attributes` to all the components so consumers can easily add attributes to them. Most (all?) of our components work this way.

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 14 above)
Comment (KristinLBradley) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:70]: I think showing one or more typical use cases would probably be the most helpful to users.

File: packages/components/NEW-COMPONENT-CHECKLIST.md (hunk 2 above)
Comment (KristinLBradley) [reply packages/components/NEW-COMPONENT-CHECKLIST.md:82]: @didoo Do you think it's helpful to include this note? I can delete it if it's not. I deleted the other text I added on branch naming since Melanie didn't seem to agree with it.

Comment (didoo) [reply packages/components/NEW-COMPONENT-CHECKLIST.md:82]: OK to leave it as is.

File: packages/components/tests/dummy/app/templates/components/button-set.hbs (hunk 5 above)
Comment (didoo) [reply packages/components/tests/dummy/app/templates/components/button-set.hbs:65]: @KristinLBradley remove the section, we can always add it if needed later

**Review by didoo — APPROVED** [review]
🚢 it!
